monkey/monkey_island/cc/setup/island_config_options.py

```diff
@@ -10,24 +12,51 @@
     DEFAULT_START_MONGO_DB,
 )
 
+_DATA_DIR = "data_dir"
+_SSL_CERT = "ssl_certificate"
+_SSL_CERT_FILE = "ssl_certificate_file"
+_SSL_CERT_KEY = "ssl_certificate_key_file"
+_MONGODB = "mongodb"
+_START_MONGODB = "start_mongodb"
+_LOG_LEVEL = "log_level"
+
 
 class IslandConfigOptions:
-    def __init__(self, config_contents: dict):
-        self.data_dir = expand_path(config_contents.get("data_dir", DEFAULT_DATA_DIR))
+    def __init__(self, config_contents: dict = None):
+        if not config_contents:
+            config_contents = {}
+        self.data_dir = expand_path(config_contents.get(_DATA_DIR, DEFAULT_DATA_DIR))
 
-        self.log_level = config_contents.get("log_level", DEFAULT_LOG_LEVEL)
+        self.log_level = config_contents.get(_LOG_LEVEL, DEFAULT_LOG_LEVEL)
 
         self.start_mongodb = config_contents.get(
-            "mongodb", {"start_mongodb": DEFAULT_START_MONGO_DB}
-        ).get("start_mongodb", DEFAULT_START_MONGO_DB)
+            _MONGODB, {_START_MONGODB: DEFAULT_START_MONGO_DB}
+        ).get(_START_MONGODB, DEFAULT_START_MONGO_DB)
 
         self.crt_path = expand_path(
-            config_contents.get("ssl_certificate", DEFAULT_CERTIFICATE_PATHS).get(
-                "ssl_certificate_file", DEFAULT_CRT_PATH
+            config_contents.get(_SSL_CERT, DEFAULT_CERTIFICATE_PATHS).get(
+                _SSL_CERT_FILE, DEFAULT_CRT_PATH
             )
         )
         self.key_path = expand_path(
-            config_contents.get("ssl_certificate", DEFAULT_CERTIFICATE_PATHS).get(
-                "ssl_certificate_key_file", DEFAULT_KEY_PATH
+            config_contents.get(_SSL_CERT, DEFAULT_CERTIFICATE_PATHS).get(
+                _SSL_CERT_KEY, DEFAULT_KEY_PATH
             )
         )
+
+    def update(self, target: dict):
+        target = self._expand_config_paths(target)
+        self.__dict__.update(target)
+
+    @staticmethod
+    def _expand_config_paths(config: dict) -> dict:
+        config_paths = [_DATA_DIR, f"{_SSL_CERT}.{_SSL_CERT_FILE}", f"{_SSL_CERT}.{_SSL_CERT_KEY}"]
+
+        for config_path in config_paths:
+            try:
+                expanded_val = expand_path(util.get(config, config_path, "."))
+                util.set(config, config_path, expanded_val, ".")
+            except KeyError:
+                pass
```

Comment:
When would there ever be a KeyError? Shouldn't the `.get()` calls avoid this?

Reply:
Key error will happen when some config value is NOT provided in the `server_config.json`. We can't create `IslandConfigOptions` from the contents of the file, because unspecified options will get defaults. Then, how will we know if user provided a value that overrides package config OR was it just a default and we shouldn't override the package config with it?

Reply:
Ah, I see. "." is the separator, not the default value.

Why don't we change the order of the calls in `update()`. Then call `self._expand_config_paths()`. This way, we'll be guaranteed that all paths exist.